#181414761 feature(user-profile):create user user profile

Procech20 opened the pull request:
### What does this PR do?
Create user profile after registration.

### Description of Task to be completed?
- Create user profile route
- Tested the route
- Documented the route

### How should this be manually tested?
To test the changes manually:
* clone repo on your local machine
* switch to develop branch  by typing the command ```git checkout ft-user-profile-181414761```
* Run: ```yarn install```: to install all dependencies
* Create a ```.env``` and add the variables as mentioned in the ```.env.example``` file.
* Open the host address using an HTTP client such as ``postman``` or any other.
* Make ``GET``  on  ``/api/users/profile`` to retrieve user's profile.
* Make ``PUT``  on  ``/api/users/profile/complete`` to complete user's profile.
* Make ``PATCH``  on  ``/api/users/profile/update`` to update user's profile.

### What to expect after running the commands:
* _On successful, A user will get a status of ``200`` for profile fetch, profile update and profile complete_

### Any background context you want to provide?
- N/A

``` Useful links ```:
- N/A

### What are the relevant pivotal tracker stories?
* https://www.pivotaltracker.com/story/show/181414761

### Screenshots (if appropriate)
- N/A

### Questions:
- N/A


---


### Task implementation plan

#1. Name of task
- Profile Page Settings

#2. Pivot Tracker ID

- [181414761](https://www.pivotaltracker.com/story/show/181414761)

#3. Branch Name

- ft-create-user-profile-181414761

#4. Description of task (Paraphrase what we have in Pivot Tracker)

* After signup on Barefoot Nomad, a user should be able to have their profile created and may update the account to add some more information.


#5. Acceptance criteria (Checks to be made)

- [ User should be able to Edit my profile, but should NOT be able to edit another user's profile]

```Scenario```: Given User registers to be a member of Barefoot Nomad.Registration is done.User Profile is created.A timestamp is associated to the profile

#6. Database schema (Migration & models)
 
<img width="431" alt="Screen Shot 2022-04-18 at 18 40 37" src="https://user-images.githubusercontent.com/75941936/163841443-2b591819-30e9-4b7d-af84-e41fa033f7e0.png">

#7. Validations point (Input to validate)
* Will use Joi validator to check:

```For profile```:

- Gender - string, valid(‘Male’, ’Female’, ’Prefer Not say’)
- Date of Birth - date, iso, 
- Gender: string,
- Age: number,
- Nationality: string,
- Preferred Language: string
- Preferred Currency: string
- Department: stringAddress: string
- Line Manager: string
- Phone Number: number
- Marital Status: string, valid(‘Single’, ‘Married’, ‘Widowed’)

#8. Logic on the access (Explaining who is allowed to CRUD (if applicable) the entities you are working on)

* User: Allowed to view and update their profile after registration

#9. Logic on actual Implementation

- Update User model for user 
- Setup the validations using Joi
- Use middleware to check if user is logged in, is profile owner
- Create controllers for profile retrieve and update
- Create the route for profile retrieve and update
- Test the routes with mocha & chai
- Document the endpoint with swagger

#10. Endpoints Format: {url}/api/user/….

GET   - /profile: Get user profile - Owner
PUT   - /profile/complete: complete user data - Owner
PATCH   - /profile/complete: update user data - Owner

Robertishimwe commented on src/database/migrations/20220417175812-modify_users_add_new_fields.js:
```diff
@@ -0,0 +1,73 @@
+    return Promise.all([
+      queryInterface.removeColumn('Users', 'age'),
+      queryInterface.removeColumn('Users', 'gender'),
+      queryInterface.removeColumn('Users', 'dob'),
```
Does this dob means data of birth? I think you don't need to have column of age and column of dob at the same time. You can remove age since we can know it from date of birth

ericus123 replied:
@Procech20  age can be filled automatically depending on the dob entered

Robertishimwe commented on src/routes/auth.js:
```diff
@@ -22,7 +22,7 @@ const router = Router();
   '/register',
   UserValidation.verifyUser,
-  EmailValidation.checkEmail,
+  // EmailValidation.checkEmail,
```
Remove this

Procech20 replied:
This is the middleware for checking valid email, it is commented because the deep email validator has issue. If resolved that middleware will be uncommented.

Robertishimwe commented on src/controllers/profile.js:
```diff
@@ -0,0 +1,189 @@
+    try {
+      const updated = await updateUser(req.user, {
+        ...req.body,
+        profileImage: '',
```
How is this profileImage linked with user? I can see you did not add it in the model or migration

ericus123 commented on src/routes/user.js:
```diff
@@ -10,6 +15,9 @@ const { validateEmail } = EmailValidation;
 const { verifyUser, inputNewPassword, setNewPassword } = UserController;
 
 const { resendConfirmationEmail, sendResetPasswordEmail } = EmailController;
+// const { verifyAuth } = profileMiddleware;
+// const { resendConfirmationEmail, verifyUser } = UserController;
```
@Procech20  why are these commented out???

Procech20 replied:
They are no longer in use anywhere

niyojuste commented on src/utils/successResponse.js:
```diff
@@ -1,3 +1,3 @@
-export default (res, status, body) => {
-  res.status(status).json(body);
+export default (res, status, message, body) => {
+  res.status(status).json({ message, body });
```
Make sure you're carefully thorough with all the handlers depending on this change. The social auth controllers, for example, are among those and you seem to have missed them.

niyojuste replied:
Make sure you're carefully thorough with all the handlers depending on this change. The social auth controllers, for example, are among those and you seem to have missed them.

itisWasp committed:
ft(user-profile): Create user profile after registration

Implementations:
 - View user profile.
 - Complete user profile
 - Update user profile after completion
 - Test all endpoints
 - Document all endpoints

[Delivers #181414741]

Robertishimwe commented on src/database/migrations/20220417175812-modify_users_add_new_fields.js:
```diff
@@ -0,0 +1,78 @@
+        allowNull: true,
+      }),
+      queryInterface.addColumn('Users', 'lineManager', {
+        type: Sequelize.STRING,
```
isn't this supposed to be integer type? i think it has to be id of line manager

Procech20 replied:
No, I haven't been updated to what the line Manager should hold.

itisWasp reviewed (changes requested):
Rebase and make sure CI is Passing.

ericus123 commented:
@Procech20 why is this not based from develop and the tests are still failing

ericus123 commented:
@Procech20 rebase
